{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a280d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "\n",
    "import sys\n",
    "sys.path.append(\"/home/dzilov/Dropbox/gepymics/\")\n",
    "from gepymics import fasta_read\n",
    "from gepymics import reverse_complement\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd4f03b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DnaA\n",
    "\n",
    "anno_files = glob(\"/mnt/projects/zilov/data/mycoplasma/annotation/*gff\")\n",
    "nucl_prot_files = glob(\"/mnt/projects/zilov/data/mycoplasma/genomes/*fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4580fb29",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnaa_ids = defaultdict()\n",
    "\n",
    "for i, file in enumerate(anno_files):\n",
    "    with open(file) as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue            \n",
    "            if \"dnaA\" in line:\n",
    "                line = line.split(\"\\t\")\n",
    "                coordinates = [int(line[3]), int(line[4]), line[6]]\n",
    "                dnaa_ids[line[0]] = coordinates\n",
    "                i += 1\n",
    "print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c0f340b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnaa_fasta = {}\n",
    "\n",
    "i = 0\n",
    "for file in nucl_prot_files:\n",
    "    fasta = fasta_read(file)\n",
    "    for nid, coordinates in dnaa_ids.items():\n",
    "        if nid in fasta.keys():\n",
    "            if coordinates[-1] == \"-\":\n",
    "                dnaa_fasta[nid] = reverse_complement(fasta[nid][coordinates[0]-1:coordinates[1]-1])\n",
    "            else:\n",
    "                dnaa_fasta[nid] = fasta[nid][coordinates[0]-1:coordinates[1]-1]\n",
    "            i += 1\n",
    "            \n",
    "print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460a2ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnaa_file = \"/media/eternus1/projects/zilov/data/mycoplasma/genes_fasta_to_msa/dnaa.fasta\"\n",
    "\n",
    "with open(dnaa_file, \"w\") as fw:\n",
    "    for h, s in dnaa_fasta.items():\n",
    "        fw.write(f\">{h}\\n{s}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a252070f",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_file = \"/media/eternus1/projects/zilov/data/mycoplasma/genes_fasta_to_msa/lcl_mod_ref-NZ_JNJG01000004.1 and 51 other sequences.aln\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdff14af",
   "metadata": {},
   "outputs": [],
   "source": [
    "longest_atgc = \"\"\n",
    "with open(consensus_file) as fh:\n",
    "    score = 0\n",
    "    max_score = 0\n",
    "    miss = 0\n",
    "    for line in fh:\n",
    "        if not line.startswith(\">\"):\n",
    "            for s in line:\n",
    "                if score > max_score:\n",
    "                    max_score = score\n",
    "                if s in [\"A\", \"T\", \"G\", \"C\"]:\n",
    "                    if score < 0:\n",
    "                        score = 0\n",
    "                    score += 1\n",
    "                    print(s, score)\n",
    "                elif s in [\"W\", \"K\", \"N\", \"R\", \"M\", \"Y\", \"S\", \"B\", \"D\", \"H\", \"V\"]:\n",
    "                    score -= 1\n",
    "                    print(s, score)\n",
    "                else: \n",
    "                    score = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14bac7f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72c4092a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_fasta = glob(\"/home/dzilov/data_zilov/data/mycoplasma/*/*/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ffac45",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "def get_organism(gff):\n",
    "    genome_dir = os.path.dirname(gff)\n",
    "    report = glob(f'{genome_dir}/*.gbff')\n",
    "    \n",
    "    with open(report[0]) as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith('SOURCE'):\n",
    "                organism = line.split('SOURCE')[-1].strip()\n",
    "    return organism\n",
    "\n",
    "def get_pid(gff, protein_name):\n",
    "    with open(gff) as fh:\n",
    "        for line in fh:\n",
    "            if protein_name.lower() in line.lower():\n",
    "                pid = line.split(\"\\t\")[8].split(\"Name=\")[-1].split(\";\")[0]\n",
    "                return pid\n",
    "                \n",
    "def fasta_read(path_to_fasta_file):\n",
    "    \"Returns dictionary header:sequence\"\n",
    "    fasta = {}\n",
    "    header = None\n",
    "    with open(path_to_fasta_file) as fh:\n",
    "        for i, line in enumerate(fh):\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if header:\n",
    "                    fasta[header] = \"\".join(seq)\n",
    "                header = line[1:].split()[0]\n",
    "                seq = []\n",
    "            else:\n",
    "                seq.append(line)\n",
    "        if header:\n",
    "            fasta[header] = \"\".join(seq)\n",
    "    return fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a7f019",
   "metadata": {},
   "outputs": [],
   "source": [
    "pids2bact = {} \n",
    "for file in anno_files:\n",
    "    if get_pid(file, 'chromosomal replication initiator protein DnaA'): \n",
    "        sco = 'chromosomal replication initiator protein DnaA'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    elif get_pid(file, 'chromosomal replication initiat'):\n",
    "        sco = 'chromosomal replication initiat'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    elif get_pid(file, 'chromosomal replication protein'):\n",
    "        sco = 'chromosomal replication protein'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    else:\n",
    "        not_found += 1\n",
    "    pids2bact[prot2bact[0]] = prot2bact[1]\n",
    "    \n",
    "print(found, not_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d84c16c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101 prots added to multifasta /home/dzilov/data_zilov/data/mycoplasma/clusters/multifasta/dnaa.fasta\n"
     ]
    }
   ],
   "source": [
    "## sco num_bact n_missed uniq_prots prots_in_multifasta multifasta_file\n",
    "summary_table = []\n",
    "\n",
    "anno_files = glob(\"/home/dzilov/data_zilov/data/mycoplasma/*/*/*.gff\")\n",
    "\n",
    "print(len(anno_files), 'genomes to analyze')\n",
    "\n",
    "# read all cds fastas\n",
    "all_fastas = []\n",
    "for i in glob(\"/home/dzilov/data_zilov/data/mycoplasma/*/*/*cds_from_genomic.fna\"):\n",
    "    fasta = fasta_read(i)\n",
    "    all_fastas.append(fasta)\n",
    "    \n",
    "# build multifasta for each sco by all bacteria\n",
    "\n",
    "\n",
    "multifasta_file = f\"/home/dzilov/data_zilov/data/mycoplasma/clusters/multifasta/dnaa.fasta\"\n",
    "\n",
    "#     if os.path.exists(multifasta_file) and os.path.getsize(multifasta_file) != 0:\n",
    "#         print(f\"Multifasta for {sco} already written, continue...\")\n",
    "#         continue\n",
    "\n",
    "\n",
    "pids2bact = {} \n",
    "n_missed = 0\n",
    "for file in anno_files:\n",
    "    if get_pid(file, 'chromosomal replication initiator protein DnaA'): \n",
    "        sco = 'chromosomal replication initiator protein DnaA'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    elif get_pid(file, 'chromosomal replication initiat'):\n",
    "        sco = 'chromosomal replication initiat'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    elif get_pid(file, 'chromosomal replication protein'):\n",
    "        sco = 'chromosomal replication protein'\n",
    "        prot2bact = [get_pid(file, sco), get_organism(file)]\n",
    "    else:\n",
    "        n_missed += 1\n",
    "        print(f\"MISSED {sco} in {file}, organism: {prot2bact[1]}\")\n",
    "    pids2bact[prot2bact[0]] = prot2bact[1]\n",
    "\n",
    "\n",
    "print(f'{len(pids2bact)} uniq prots in {sco}')\n",
    "\n",
    "multifasta = {}\n",
    "for pid, organism in pids2bact.items():\n",
    "    for prots in all_fastas:\n",
    "        for header, seq in prots.items():\n",
    "            if pid in header:\n",
    "                short_organism = \"_\".join(organism.split(\",\")[0].split())\n",
    "                multifasta[f\"{pid}_{short_organism}\"]=seq\n",
    "\n",
    "\n",
    "\n",
    "with open(multifasta_file, \"w\") as fw:\n",
    "    for h, s in multifasta.items():\n",
    "        fw.write(f\">{h}\\n{s}\\n\")\n",
    "\n",
    "print(len(multifasta.keys()), f'prots added to multifasta {multifasta_file}')\n",
    "\n",
    "sco_summary = [sco, str(len(all_fastas)), str(n_missed), str(len(pids2bact)), multifasta_file]\n",
    "summary_table.append(sco_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f326148",
   "metadata": {},
   "outputs": [],
   "source": [
    "def msa_to_consensus(fasta_msa, only_n_consensus=False, compare_treshold=.3,):\n",
    "\n",
    "    consensus_table = defaultdict()\n",
    "    \n",
    "    fasta_msa = fasta_read(fasta_msa)\n",
    "\n",
    "    for v in fasta_msa.values():\n",
    "        for i in range(len(v)):\n",
    "            consensus_table[i] = {'A': 0, 'C': 0, 'G': 0, 'T': 0, '-': 0, 'N': 0}\n",
    "        break\n",
    "    \n",
    "    for k, v in fasta_msa.items():\n",
    "        for i, n in enumerate(v):\n",
    "            consensus_table[i][n] += 1\n",
    "\n",
    "    consensus = []\n",
    "\n",
    "    for k, v in consensus_table.items(): \n",
    "        few_n = []\n",
    "\n",
    "        for n, c in v.items():\n",
    "            if c == len(fasta_msa):\n",
    "                few_n.append(n)\n",
    "            else:\n",
    "                if c != 0:\n",
    "                    if n == '-':\n",
    "                        if c / len(fasta_msa) > 0.7:\n",
    "                            few_n = []\n",
    "                    else:\n",
    "                        if c / len(fasta_msa) > compare_treshold:\n",
    "                            few_n.append(n)\n",
    "                            \n",
    "        if len(few_n) == 0:\n",
    "            continue\n",
    "        elif len(few_n) == 1:\n",
    "            if '-' in few_n or 'N' in few_n:\n",
    "                consensus.append('N')\n",
    "            else:\n",
    "                consensus.append(few_n[0])\n",
    "        else:\n",
    "            if only_n_consensus:\n",
    "                consensus.append('N')\n",
    "            else:\n",
    "                if ('A' in few_n and 'T' in few_n and 'G' in few_n and 'C' in few_n) or ('-' in few_n or 'N' in few_n):\n",
    "                    consensus.append('N')\n",
    "                elif 'A' in few_n and 'G' in few_n:\n",
    "                    if 'C' in few_n:\n",
    "                        consensus.append('V')\n",
    "                    elif 'T' in few_n:\n",
    "                        consensus.append('D')\n",
    "                    else:\n",
    "                        consensus.append('R')\n",
    "                elif 'A' in few_n and 'T' in few_n:\n",
    "                    if 'C' in few_n:\n",
    "                        consensus.append('H')\n",
    "                    else:\n",
    "                        consensus.append('W')\n",
    "                elif 'G' in few_n and 'C' in few_n:\n",
    "                    if 'T' in few_n:\n",
    "                        consensus.append('B')\n",
    "                    else:\n",
    "                        consensus.append('S')\n",
    "                elif 'G' in few_n and 'T' in few_n:\n",
    "                    consensus.append('K')\n",
    "                elif 'C' in few_n and 'T' in few_n:\n",
    "                    consensus.append('Y')\n",
    "                elif 'A' in few_n and 'C' in few_n:\n",
    "                    consensus.append('M')\n",
    "\n",
    "    return ''.join(consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4580d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df482bf4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f7ee0da",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
